src/v/cluster/controller_backend.cc

```diff
@@ -292,29 +292,32 @@ void controller_backend::housekeeping() {
 }
 
 ss::future<> controller_backend::reconcile_ntp(deltas_t& deltas) {
-    return ss::do_with(
-      bool{false},
-      deltas.cbegin(),
-      [this, &deltas](bool& stop, deltas_t::const_iterator& it) {
-          return ss::do_until(
-                   [&stop, &it, &deltas] {
-                       return stop || it == deltas.cend();
-                   },
-                   [this, &it, &stop] {
-                       return execute_partitition_op(*it).then(
-                         [&it, &stop](std::error_code ec) {
-                             if (ec) {
-                                 stop = true;
-                                 return;
-                             }
-                             it++;
-                         });
-                   })
-            .then([&deltas, &it] {
-                // remove finished tasks
-                deltas.erase(deltas.cbegin(), it);
-            });
-      });
+    bool stop = false;
+    auto it = deltas.begin();
+    while (!(stop || it == deltas.end())) {
+        try {
+            auto ec = co_await execute_partitition_op(*it);
+            if (ec) {
+                vlog(
+                  clusterlog.info,
+                  "partition operation error: {} - {}",
+                  *it,
+                  ec.message());
+                stop = true;
+                continue;
+            }
+        } catch (...) {
+            vlog(
+              clusterlog.error,
+              "exception while executing partiton operation: {} - {}",
+              *it,
+              std::current_exception());
+            stop = true;
+            continue;
+        }
+        ++it;
+    }
+    deltas.erase(deltas.begin(), it);
```

Comment:
in pre-coroutine code it looks like the erase doesn't occur if execute_partitition_op returns an exceptional future, but now erase always occurs.

Reply:
it is safe to erase since when we advance `it` we are sure operation succeed

Reply:
in case `it` is `begin()` erase would be a nop

Reply:
I mean before if any `execute_partitition_op` failed no erases would occur. That is no longer the case. If that is ok then we can merge this.